service/qos/service_level_controller.cc

```diff
@@ -307,18 +328,24 @@ future<> service_level_controller::notify_service_level_removed(sstring name) {
     co_return;
 }
 
-void service_level_controller::update_from_distributed_data(std::function<steady_clock_type::duration()> interval_f) {
+void service_level_controller::update_from_distributed_data(std::function<steady_clock_type::duration()> interval_f, service::storage_service& storage_service, service::raft_group0_client& group0_client) {
     if (this_shard_id() != global_controller) {
         throw std::runtime_error(format("Service level updates from distributed data can only be activated on shard {}", global_controller));
     }
     if (_global_controller_db->distributed_data_update.available()) {
         sl_logger.info("update_from_distributed_data: starting configuration polling loop");
         _logged_intervals = 0;
-        _global_controller_db->distributed_data_update = repeat([this, interval_f = std::move(interval_f)] {
+        _global_controller_db->distributed_data_update = repeat([this, interval_f = std::move(interval_f), &group0_client, &storage_service] {
             return sleep_abortable<steady_clock_type>(interval_f(),
-                    _global_controller_db->dist_data_update_aborter).then_wrapped([this] (future<>&& f) {
+                    _global_controller_db->dist_data_update_aborter).then_wrapped([this, &group0_client, &storage_service] (future<>&& f) {
                 try {
                     f.get();
+                    
+                    if (!group0_client.in_recovery() && storage_service.get_topology_upgrade_state() == service::topology::upgrade_state_type::done) {
+                        stop_update_from_distributed_data();
+                        return make_ready_future<stop_iteration>(stop_iteration::yes);
```

Comment:
I think that we should print some message here. The message should mention that the polling loop is being stopped, but also explain that service level updates will now be immediate and the loop will no longer be necessary.